cmd/soroban-rpc/internal/methods/get_events.go

```diff
@@ -0,0 +1,475 @@
+	*toid.ID
+	EventOrder int32
+}
+
+// String returns a string representation of this id
+func (id EventID) String() string {
+	return fmt.Sprintf(
+		"%019d-%010d",
+		id.ToInt64(),
+		id.EventOrder+1,
+	)
+}
+
+func (id *EventID) Parse(input string) error {
+	parts := strings.SplitN(input, "-", 2)
+	if len(parts) != 2 {
+		return fmt.Errorf("invalid event id %s", input)
+	}
+
+	// Parse the first part (toid)
+	idInt, err := strconv.ParseInt(parts[0], 10, 64)
+	if err != nil {
+		return errors.Wrapf(err, "invalid event id %s", input)
+	}
+	parsed := toid.Parse(idInt)
+	id.ID = &parsed
+
+	// Parse the second part (event order)
+	eventOrder, err := strconv.ParseInt(parts[1], 10, 64)
+	if err != nil {
+		return errors.Wrapf(err, "invalid event id %s", input)
+	}
+	// Subtract one to go from the id to the
+	id.EventOrder = int32(eventOrder) - 1
+
+	return nil
+}
+
+func (a EventStore) GetEvents(request GetEventsRequest) ([]EventInfo, error) {
+	if err := request.Valid(); err != nil {
+		return nil, err
+	}
+
+	finish := toid.AfterLedger(request.EndLedger)
+
+	var results []EventInfo
+
+	// TODO: Use a more efficient backend here. For now, we stream all ledgers in
+	// the range from horizon, and filter them. This sucks.
+	cursor := EventID{
+		ID:         toid.New(request.StartLedger, int32(0), 0),
+		EventOrder: 0,
+	}
+	if request.Pagination != nil && request.Pagination.Cursor != "" {
+		if err := cursor.Parse(request.Pagination.Cursor); err != nil {
+			return nil, errors.Wrap(err, "invalid pagination cursor")
+		}
+	}
+	err := a.ForEachTransaction(cursor.ID, finish, func(transaction horizon.Transaction) error {
+		// parse the txn paging-token, to get the transactionIndex
+		pagingTokenInt, err := strconv.ParseInt(transaction.PagingToken(), 10, 64)
+		if err != nil {
+			return errors.Wrapf(err, "invalid paging token %s", transaction.PagingToken())
+		}
+		pagingToken := toid.Parse(pagingTokenInt)
+
+		// For the first txn, we might have to skip some events to get the first
+		// after the cursor.
+		operationCursor := cursor.OperationOrder
+		eventCursor := cursor.EventOrder
+		cursor.OperationOrder = 0
+		cursor.EventOrder = 0
+		if pagingToken.ToInt64() > cursor.ToInt64() {
+			// This transaction is after the cursor, so we need to reset the cursor
+			operationCursor = 0
+			eventCursor = 0
+		}
+
+		var meta xdr.TransactionMeta
+		if err := xdr.SafeUnmarshalBase64(transaction.ResultMetaXdr, &meta); err != nil {
+			// Invalid meta back. Eek!
+			return err
+		}
+
+		v3, ok := meta.GetV3()
+		if !ok {
+			return nil
+		}
+
+		ledger := transaction.Ledger
+		ledgerClosedAt := transaction.LedgerCloseTime.Format(time.RFC3339)
+
+		for operationIndex, operationEvents := range v3.Events {
+			if int32(operationIndex) < operationCursor {
+				continue
+			}
+			for eventIndex, event := range operationEvents.Events {
+				if int32(eventIndex) < eventCursor {
+					continue
+				}
+				if request.Matches(event) {
+					v0 := event.Body.MustV0()
```

Comment:
does this result in `panic` .. is it acceptable to allow this potentially crash out the overall service or checking for type/null instead and return err?